---
title: "JEM Figure 5"
author: "Maksym Zarodniuk"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output: html_document
---

```{r setup, warning=F, message=F}
library(Seurat)
library(RColorBrewer)
library(tidyverse)
library(MetBrewer)
library(cowplot)
library(patchwork)
library(DescTools)
library(scales)
library(ggrepel)
library(ggpubr)
library(ComplexHeatmap)
library(circlize)
library(msigdbr)
library(clusterProfiler)
library(enrichplot)
library(org.Rn.eg.db)
library(GOSemSim)
library(ggtree)
library(ggdendro)
library(viridis)
library(aplot)
library(ggtext)
source("util.R")
```

```{r}
remote_dir <- "/Volumes/TOSHIBA/molpat/APECED_rats/analysis/single_cell/sc_rat_spleen/data/"
seurat_obj <- readRDS(paste0(remote_dir, "integrated_seurat_object.rds"))
```

## 5A

```{r}
col_vec <- colorRampPalette(brewer.pal(9, "Set1"))(21)[1:18]
color_mapping <- data.frame(Cell.type = levels(seurat_obj$cell.type),
                            Color = col_vec)

umap_plot <- DimPlot(seurat_obj,
        label = TRUE,
        label.size = 5,
        cols = col_vec,
        group.by = "cell.type.idx") +
  theme(legend.position = "none") +
  labs(subtitle = NULL, title = NULL)

legend <- DimPlot(seurat_obj,
                group.by = "legend",
                cols = col_vec,
                label = TRUE)

legend <- plot_grid(get_legend(legend))

umap_plot + legend
```

## 5B

```{r}
ego_s <- read_tsv("data/Figure_5/GO_simplified.tsv")
ego_s <- as_tibble(ego_s)
ego_s$GeneRatio <- sapply(ego_s$GeneRatio, function(x) {eval(parse(text=x))})
ego_s <- mutate(ego_s, log10p = -log10(p.adjust), DescriptionTrunc = StrTrunc(Description, maxlen = 50, ellipsis = "...", wbound = FALSE))

P_B <- ggplot(head(ego_s, 15),
       aes(x = GeneRatio, y = reorder(DescriptionTrunc, GeneRatio), 
                        color = `log10p`, size = Count)) + 
  geom_point() +
  geom_point(aes(size=Count), shape = 21, colour="black", stroke=0.5) +
  scale_size(range = c(0, 4)) +
  scale_colour_gradient2(
    low = ("white"),
    mid = ("white"),
    high = scales::muted("red"),
    midpoint = -log10(0.05)
  ) +
  ylab("") + 
  xlab("GeneRatio") + 
  theme_cowplot() +
  coord_cartesian(xlim=c(0, 0.2)) +
  scale_y_discrete(position = "right") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

P_B
```

## 5C

```{r}
de.results <- read_tsv("data/Figure_5/benoist_DE_results.tsv")
tonic_effect_processed <- read_tsv("data/Figure_5/tonic_effect_processed.txt")
de.results <- de.results %>%
  left_join(tonic_effect_processed %>% dplyr::select(ortholog_name, Group) %>% distinct(), by = c("Gene.name" = "ortholog_name")) %>%
  mutate(DE = if_else(FDR < .05 & logFC > 1, "Upregulated", "Not changed"),
             DE = if_else(FDR < .05 & logFC < -1, "Downregulated", DE)) %>%
  replace_na(list(Group = "Unknown"))

B_ORA <- fisher.test(as.numeric(de.results[de.results$Cell.type=="B",]$DE=="Downregulated"), as.numeric(de.results[de.results$Cell.type=="B",]$Group=="Tonic-sensitive"))
MF_ORA <- fisher.test(as.numeric(de.results[de.results$Cell.type=="MF",]$DE=="Downregulated"), as.numeric(de.results[de.results$Cell.type=="MF",]$Group=="Tonic-sensitive"))

p1 <- plot_tonic_ISGs(de.results[de.results$Cell.type=="B",]) + ggtitle("B cell") + annotate("text", x = -5.5, y = 7.2, hjust=0, col="black", label = paste0("OR=",round(B_ORA$estimate, 1),"\np=",format(B_ORA$p.value, digits = 2)))
p2 <- plot_tonic_ISGs(de.results[de.results$Cell.type=="MF",]) + ggtitle("Macrophage") + annotate("text", x = -5.5, y = 7.2, hjust=0, col="black", label = paste0("OR=",round(MF_ORA$estimate, 1),"\np=",format(MF_ORA$p.value, digits = 2)))

P_C <- p1 + p2

P_C
```

## 5D

```{r}
tonic_effect_processed <- read_tsv("data/Figure_5/tonic_effect_processed.txt")
cell_type <- seurat_obj$cell.type
levels(cell_type) <- c("T4",  "T4", "T8", "T8", "NK", "NK", "B", "B", "B", "B", "B", "B",  "MF", "MF", "MF", "DC", "DC", "GN")
seurat_obj$cell.type.lvl1 <- cell_type

celltype_vec <- levels(seurat_obj$cell.type.lvl1)
seurat_obj$celltype.gntp <- paste(seurat_obj$cell.type.lvl1, seurat_obj$genotype, sep = "_")
seurat_obj$celltype.gntp <- ordered(seurat_obj$celltype.gntp, levels = paste(rep(celltype_vec, each = 2), rep(c("HE", "KO"), length(celltype_vec)), sep = "_"))
Idents(seurat_obj) <- "celltype.gntp"

DefaultAssay(seurat_obj) <- "RNA"
dge_df = data.frame()
for (i in unique(celltype_vec)) {
  dge_df_sub <- FoldChange(seurat_obj, ident.1 = paste(i, "_KO", sep=""), ident.2 = paste(i, "_HE", sep=""), verbose = TRUE) %>%
    rownames_to_column("gene") %>%
    mutate(Cell.type = i)
  dge_df <- rbind(dge_df, dge_df_sub)
}

mat <- dge_df %>%
  dplyr::select(-c(pct.1, pct.2)) %>%
  pivot_wider(names_from = "Cell.type", values_from = "avg_log2FC") %>%
  column_to_rownames("gene")
```

### B

```{r}
data_to_plot <- tonic_effect_processed %>%
  filter(Cell.type=="B") %>%
  dplyr::select(ortholog_name, IFNAR_KO.FC, Group) %>%
  drop_na(ortholog_name) %>%
  left_join(dge_df %>% filter(Cell.type == "B") %>% dplyr::select(gene, avg_log2FC), by=c("ortholog_name"="gene")) %>%
  mutate(avg_FC = 2^avg_log2FC, IFNAR_KO.FC = 1/IFNAR_KO.FC) %>%
  drop_na(avg_log2FC)

col_vec <- c("#E41A1C",
               "#d3d3d3",
               "#377EB8")
names(col_vec) <- c("Tonic-sensitive", "NA", "Tonic-insensitive")
lab_df <- filter(data_to_plot, Group == "Tonic-sensitive")
cor_test <- cor.test(data_to_plot$IFNAR_KO.FC, data_to_plot$avg_FC, method = "spearman")

p_D1 <- ggplot(data_to_plot, aes(x=IFNAR_KO.FC, y=avg_FC, color=Group, label = ortholog_name)) +
  geom_point(size=1) +
  theme_cowplot(16) +
  geom_hline(yintercept = 1, lty = 2) +
  geom_vline(xintercept = 1, lty = 2) +
  labs(x = "FC (Ifnar1 KO / Ctrl)", y = "FC (Aire KO / Ctrl)") +
  geom_text_repel(data = lab_df,
                  aes(x = IFNAR_KO.FC, y = avg_FC, label = ortholog_name),
                  min.segment.length = 1, seed = 3,
                  box.padding = 1,
                  show.legend = FALSE,
                  max.overlaps =33,
                  force_pull=2,
                  force=100,
                  size=5
  ) +
  annotate("text", x = 0.1, y = 5, hjust = 0, 
           label = paste0("rho=", round(cor_test$estimate, 2), 
                          "\n", "p=", format(cor_test$p.value, digits=3))) +
  scale_color_manual(values = col_vec) +
  scale_x_log10() +
  scale_y_log10() +
  theme(legend.position = "none") +
  ggtitle("B cell")
```

### MF

```{r}
data_to_plot <- tonic_effect_processed %>%
  filter(Cell.type=="MF") %>%
  dplyr::select(ortholog_name, IFNAR_KO.FC, Group) %>%
  drop_na(ortholog_name) %>%
  left_join(dge_df %>% filter(Cell.type == "MF") %>% dplyr::select(gene, avg_log2FC), by=c("ortholog_name"="gene")) %>%
  mutate(avg_FC = 2^avg_log2FC, IFNAR_KO.FC = 1/IFNAR_KO.FC) %>%
  drop_na(avg_log2FC)

col_vec <- c("#E41A1C",
               "#d3d3d3",
               "#377EB8")
names(col_vec) <- c("Tonic-sensitive", "NA", "Tonic-insensitive")
lab_df <- filter(data_to_plot, Group == "Tonic-sensitive")
cor_test <- cor.test(data_to_plot$IFNAR_KO.FC, data_to_plot$avg_FC, method = "spearman")
  
p_D2 <- ggplot(data_to_plot, aes(x=IFNAR_KO.FC, y=avg_FC, color=Group, label = ortholog_name)) +
  geom_point(size=1) +
  theme_cowplot(16) +
  geom_hline(yintercept = 1, lty = 2) +
  geom_vline(xintercept = 1, lty = 2) +
  labs(x = "FC (Ifnar1 KO / Ctrl)", y = "FC (Aire KO / Ctrl)") +
  geom_text_repel(data = lab_df,
                  aes(x = IFNAR_KO.FC, y = avg_FC, label = ortholog_name),
                  min.segment.length = 1, seed = 10,
                  box.padding = 1,
                  show.legend = FALSE,
                  max.overlaps =25,
                  force_pull=100,
                  force=100,
                  nudge_x=.2,
                  size=5
  ) +
  annotate("text", x = 0.05, y = 10, hjust = 0, 
           label = paste0("rho=", round(cor_test$estimate, 2), 
                          "\n", "p=", format(cor_test$p.value, digits=3))) +
  scale_color_manual(values = col_vec) +
  scale_x_log10() +
  scale_y_log10() +
  theme(legend.position = "none") +
  ggtitle("Macrophage")
```

```{r}
p_D <- p_D1 + p_D2

p_D
```

## 5E

```{r}
tonic_effect <- read_tsv("data/Figure_5/tonic_effect_processed.txt")
de_data <- read_tsv("data/Figure_5/Tcells_DE_data.tsv")
exprs_mtx <- read_tsv("data/Figure_5/exprs_mtx.txt")
exprs_mtx$SYMBOL <- NULL

de_data <- de_data %>%
  filter(Site != "Pooled", ISG == 1) %>%
  dplyr::select(SYMBOL, Log2FC, Site, Cell.type) %>%
  mutate(Group = paste(Site, Cell.type, sep = "_")) %>%
  dplyr::select(-c(Site, Cell.type)) %>%
  distinct(SYMBOL, Group, .keep_all = T) %>%
  pivot_wider(names_from = Group, values_from = Log2FC) %>%
  column_to_rownames("SYMBOL") %>%
  drop_na() %>%
  as.matrix()
  
col_names <- colnames(de_data)
colnames(de_data) <- str_split(col_names, "_", Inf, T)[,2]
tonic_effect_sub <- tonic_effect[match(rownames(de_data), tonic_effect$ortholog_name),] %>%
  replace_na(list(Group = "No data"))

set.seed(42)
cl <- kmeans(de_data, centers = 3)$cluster
colAnn <- HeatmapAnnotation(Tonicity = tonic_effect_sub$Group,
                            which = 'col',
                            col = list(Tonicity = c("Tonic-sensitive" = "#E41A1C", "Tonic-insensitive" = "#377EB8", "No data"="white")),
                            simple_anno_size = unit(.5, "cm"),
                            gap = unit(1, 'cm'),
                            show_annotation_name = T)

P_E = Heatmap(t(de_data),
        name = "log2FC",
        #col = colorRamp2(seq(-3, 3, length.out=9), rev(brewer.pal(n=9, name="Spectral"))),
        row_split = factor(str_split(col_names, "_", Inf, T)[,1]),
        column_split = cl,
        border=T,
        show_row_dend = F,
        show_column_dend = F,
        show_column_names = F,
        top_annotation = colAnn)

P_E
```

## 5F

```{r, fig.width=15, fig.height=7}
de_data <- read_tsv("data/Figure_5/Tcells_DE_data.tsv")
go.result_df <- read_tsv("data/Figure_5/Tcells_GO_full.tsv")
go.result.simpl_df <- read_tsv("data/Figure_5/Tcells_GO_simplified.tsv")

de_data$SYMBOL <- str_split(de_data$SYMBOL, "; ", Inf, T)[,1]
list_names <- unique(de_data$Cell.type)
de_data_list <- group_split(de_data, Cell.type)
de_data_list <- as.list(de_data_list)
names(de_data_list) <- list_names

hsGO <- godata('org.Rn.eg.db', ont="BP")
go.result.simpl_df$GeneRatioDecimal <- sapply(go.result.simpl_df$GeneRatio, function(x) {eval(parse(text=x))})

data_to_plot <- go.result.simpl_df %>%
  group_by(Cell.type) %>%
  slice_min(order_by = p.adjust, n = 25)
data_to_plot <- go.result.simpl_df[go.result.simpl_df$ID %in% go.result.simpl_df$ID,]
data_to_plot <- data_to_plot %>%
  mutate(log10p = -log10(p.adjust),
         DescriptionTrunc = StrTrunc(Description, maxlen = 60, ellipsis = "...", wbound = FALSE))

sim_mtx <- mgoSim(unique(data_to_plot$ID), unique(data_to_plot$ID), semData=hsGO, measure="Wang", combine=NULL)
dist_mtx <- 1/sim_mtx
col_clust <- hclust(as.dist(dist_mtx))
ddgram_col <- as.dendrogram(col_clust)
ggtree_plot_col <- ggtree(ddgram_col)

sc <- scale_colour_gradient2(
    low = "white",
    mid = "white",
    high = scales::muted("red")
  )

data_to_plot$ID <- factor(data_to_plot$ID, levels=col_clust$labels[col_clust$order])

dotplot <- ggplot(data_to_plot, aes(x=Cell.type, y = DescriptionTrunc, color = log10p, size = GeneRatioDecimal)) + 
  geom_point() +
  geom_point(aes(size=GeneRatioDecimal), shape = 21, colour="black", stroke=0.5) +
  cowplot::theme_cowplot() +
  theme(axis.text.x = element_text(angle = 90,  vjust = 0.5, hjust=1),
        axis.line = element_blank()) +
  ylab('') +
  theme(axis.ticks = element_blank()) +
  scale_y_discrete(position = "right") +
  sc

ggtree_plot_col <- ggtree_plot_col + aplot::ylim2(dotplot)

P_F <- ggtree_plot_col +
  dotplot + 
  plot_layout(ncol = 2, widths = c(0.7, 4))

P_F
```

## S3

```{r, fig.height=15, fig.width=7}
markers_df <- read_tsv("data/Figure_5/marker_genes.txt")
color_mapping <- colorRampPalette(brewer.pal(9, "Set1"))(21)[1:18]
names(color_mapping) <- levels(seurat_obj$cell.type)
markers_df$Color <- color_mapping[match(markers_df$gene.corresp, names(color_mapping))]
markers_df <- markers_df %>%
  replace_na(list(Color = "#b1b1b1")) %>%
  mutate(y.label = paste("<span style = 'color:",
                         Color,
                         ";'>",
                         gene,
                         "</span>", sep = ""))

data.to.plot <- dot_plot(object = seurat_obj[,seurat_obj$genotype=="HE"],
        assay = "RNA",
        features = markers_df$gene,
        cluster.idents = TRUE,
        group.by = "cell.type") %>%
  dplyr::rename(`% Expressing` = pct.exp, `Cell type` = id) %>%
  left_join(markers_df, by = c("features.plot" = "gene"))

data.to.plot$features.plot <- factor(data.to.plot$features.plot, levels = unique(markers_df$gene))

data.to.plot <- data.to.plot %>%
  filter(`% Expressing` > 1)
  
data.to.plot$y.label <- factor(data.to.plot$y.label, levels = markers_df$y.label)
data.to.plot$`Cell type` <- factor(data.to.plot$`Cell type`, levels = (levels(seurat_obj$cell.type)))

dotplot <- data.to.plot %>%
  ggplot(aes(y=features.plot, x = `Cell type`, color = avg.exp.scaled, size = `% Expressing`)) + 
  geom_point() +
  geom_point(aes(size=`% Expressing`), shape = 21, colour="black", stroke=0.5) +
  theme_cowplot(13) +
  scale_radius(limits = c(0, 100)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
        axis.text.y = element_markdown(),
        axis.line = element_blank()) +
  ylab(NULL) +
  xlab(NULL) +
  theme(axis.ticks = element_blank()) +
  scale_x_discrete(position = "top") +
  viridis::scale_colour_viridis(option="magma") +
  guides(size=guide_legend(override.aes=list(shape=21, colour="black", fill="white")))

labels <- ggplot(data.to.plot %>%
                   distinct(`Cell type`),
                 aes(y = factor(1), x = `Cell type`, fill = `Cell type`)) + 
  geom_tile() +
  scale_fill_manual(values = color_mapping) + 
  theme_nothing() +
  xlim2(dotplot) +
  xlab(NULL)

P <- labels +
  dotplot + 
  plot_layout(nrow = 2, heights = c(.2, 9))

P
```

## S4A

```{r}
# This ISG list is from Interferome DB
rat_isgs <- read.table("data/Figure_5/rat_isgs.tsv", header = TRUE, sep = "\t")

# DE analysis genes
de.results <- read.table("data/Figure_5/DE_results.tsv", header = TRUE, sep = "\t")

# Preselecting cell types
selected_cell_types <- de.results %>% filter(FDR < 0.05) %>% group_by(Cell.type) %>% summarise(Count = n()) %>% pull(Cell.type)

# Enrichment of ISGs in downregulated genes
enrichment_df = data.frame()
for (Cell_type in selected_cell_types) {
  de.results_sub <- de.results %>%
    dplyr::filter(Cell.type == Cell_type) %>%
    mutate(DE = ifelse(FDR < 0.05 & logFC < 1, 1, 0)) %>%
    dplyr::select(Gene.name, DE)
  
  rat_de_isg <- de.results_sub %>%
    left_join(rat_isgs, by=c("Gene.name"="Rat_symbol")) %>%
    replace_na(list(ISG = 0))
  
  contingency.table <- table(rat_de_isg$ISG, rat_de_isg$DE)
  
  if (ncol(contingency.table) < 2) {
    enrichment_df <- rbind(
      enrichment_df,
      data.frame(Cell.type = Cell_type,
                 pval = NA,
                 OR = NA,
                 CI.low = NA,
                 CI.hi = NA))
    next
  }
  
  test <- fisher.test(contingency.table)
  
  enrichment_df <- rbind(
    enrichment_df,
    data.frame(Cell.type = Cell_type,
               pval = test$p.value,
               OR = test$estimate,
               CI.low = test$conf.int[1],
               CI.hi = test$conf.int[2])
  )
}

enrichment_df$p.adj <- p.adjust(enrichment_df$pval)

enrichment_df <- enrichment_df %>%
  mutate(Col = if_else(p.adj < .05, "p<0.05", "p>0.05"))

col_vec <- as.vector(met.brewer("Nizami", 5)[4])

enrichment_df <- enrichment_df %>%
  mutate(hjust = ifelse(CI.hi==Inf, 1.2, 1.2),
         hjust = ifelse(is.na(CI.hi), NA, hjust))

enrichment_df$Cell.type <- factor(enrichment_df$Cell.type, levels = enrichment_df$Cell.type[order(enrichment_df$OR)])

P_S4A <- ggplot(enrichment_df, aes(x = Cell.type, y = OR, ymin = CI.low, ymax = CI.hi, color=Col)) + 
  geom_linerange() + 
  geom_pointrange() +
  geom_text(aes(label = formatC(p.adj, format = "e")), hjust=enrichment_df$hjust, vjust=-.4) +
  scale_y_log10() +
  coord_flip() +
  #geom_hline(yintercept = 1, linetype = "dashed") +
  theme_cowplot(16) +
  #ylim(0, 700) + 
  scale_color_manual(values = c(as.vector(met.brewer("Nizami", 5)[4]),
             "grey")) +
  ylab("OR") +
  xlab(NULL) +
  theme(legend.position = "none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

P_S4A
```

## S3

```{r}
gntp_col <- met.brewer("Nizami", 2)[c(2, 1)]
names(gntp_col) <- c("HE", "KO")

P_S3 <- seurat_obj@meta.data %>%
  group_by(genotype, seq_folder, cell.type) %>%
  summarise(Count = n()) %>%
  mutate(Freq = Count/sum(Count)) %>%
  ggplot(aes(cell.type, Freq, fill = genotype)) +
  stat_summary(geom = "bar", fun = mean, position = "dodge") +
  stat_summary(geom = "errorbar", fun.data = mean_se, position = "dodge") +
  geom_point(position = position_dodge(0.8), alpha=0.3, size = 0.5) +
  scale_fill_manual(values = gntp_col) +
  theme_cowplot() +
  theme(axis.text.x = element_text(angle = 90,  vjust = 0.5, hjust=1)) +
  labs(x = NULL, y = "Frequency")

P_S3
```

## S4B

```{r, fig.height=15, fig.width=15}
p1 <- plot_scatter("Activated NK cell")
p2 <- plot_scatter("Macrophage (M1-like)")
p3 <- plot_scatter("Neutrophil")
p4 <- plot_scatter("Naive CD4+ T cell")
p5 <- plot_scatter("pDC")
p6 <- plot_scatter("NK cell")
p7 <- plot_scatter("FO B cell")
p8 <- plot_scatter("CD8+ Tcm cell")
p9 <- plot_scatter("Memory B cell 2")

P_S4B <- p1 + p2 + p3 + p4 + p5 + p6 + p7 + p8 + p9 + plot_layout(ncol = 3)
P_S4B
```

```{r}
sessionInfo()
```

